### PR TITLE
Refactor edit reservation unit

### PR DESCRIPTION
**Pull request by joonatank**
Replace edit page redux with react-hook-form.
Replace Joi with zod.
Improve form validators.
Add components for Accordion and TimeInput that can be controlled from outside.
Move edit page into spa directory instead of components.

Refs: TILA-2832, TILA-2830, TILA-2816, TILA-2833

### CODE REVIEW

File: apps/admin-ui/src/component/Accordion.tsx [hunk 1]
```diff
@@ -75,19 +77,40 @@ const ToggleIconClosed = styled(IconAngleDown).attrs({
...
+  initiallyOpen = false,
+  open,
   children,
   disabled = false,
   className,
   style,
   ...rest
 }: IProps): JSX.Element {
-  const [isAccordionOpen, toggleOpenState] = useState(defaultOpen);
+  const [accordianOpenState, toggleOpenState] = useState(initiallyOpen);
```
Comment (vincit-matu) [inline apps/admin-ui/src/component/Accordion.tsx:105]: accordian => accordion

File: apps/admin-ui/src/spa/ReservationUnit/edit/index.tsx [hunk 2]
```diff
@@ -0,0 +1,2311 @@
...
+          if (formValues.pk === 0 && pk > 0) {
+            history(`/unit/${unitPk}/reservationUnit/edit/${pk}`);
+          }
+          const tkey =
+            formValues.pk === 0
+              ? "ReservationUnitEditor.reservationUnitUpdatedNotification"
+              : "ReservationUnitEditor.reservationUnitCreatedNotification";
+          notifySuccess(t(tkey, { name: getValues("nameFi") }));
+        } else {
+          // FIXME error
```
Comment (joonatank) [inline apps/admin-ui/src/spa/ReservationUnit/edit/index.tsx:2048]: ??

File: apps/admin-ui/src/spa/ReservationUnit/edit/index.tsx [hunk 3]
```diff
@@ -0,0 +1,2311 @@
...
+          notifyError("failed to save images");
+          return undefined;
+        }
+      } else {
+        // FIXME error
+        notifyError("ei tullut pk");
```
Comment (joonatank) [inline apps/admin-ui/src/spa/ReservationUnit/edit/index.tsx:2054]: ???

File: apps/admin-ui/src/spa/ReservationUnit/edit/index.tsx [hunk 4]
```diff
@@ -0,0 +1,2311 @@
...
+        return undefined;
+      }
+
+      // TODO draft / archive / publish
+      notifySuccess(t("ReservationUnitEditor.saveSuccess"));
```
Comment (joonatank) [inline apps/admin-ui/src/spa/ReservationUnit/edit/index.tsx:2059]: Why double this up? there is one above.

File: apps/admin-ui/src/spa/ReservationUnit/edit/index.tsx [hunk 5]
```diff
@@ -0,0 +1,2311 @@
...
+          }
+        >
+          {t("common.prev")}
+        </WhiteButton>
+        <Preview
+          target="_blank"
+          rel="noopener noreferrer"
+          disabled={isSaving}
+          href={`${previewUrlPrefix}/${reservationUnit?.pk}?ru=${reservationUnit?.uuid}`}
+          // TODO
```
Comment (joonatank) [inline apps/admin-ui/src/spa/ReservationUnit/edit/index.tsx:2189]: ???

File: apps/admin-ui/src/spa/ReservationUnit/edit/index.tsx [hunk 6]
```diff
@@ -0,0 +1,2311 @@
...
+              ? "ReservationUnitEditor.noPreviewUnsavedChangesTooltip"
+              : "ReservationUnitEditor.previewTooltip"
+          )}
+        >
+          <span>{t("ReservationUnitEditor.preview")}</span>
+        </Preview>
+        <WhiteButton
+          size="small"
+          variant="secondary"
+          // disabled={isSaving || !hasChanges}
```
Comment (joonatank) [inline apps/admin-ui/src/spa/ReservationUnit/edit/index.tsx:2202]: ???

File: apps/admin-ui/src/spa/ReservationUnit/edit/index.tsx [hunk 7]
```diff
@@ -0,0 +1,2311 @@
...
+          isLoading={isSaving && watch("isDraft")}
+          type="button"
+          loadingText={t("ReservationUnitEditor.saving")}
+          onClick={handleSaveAsDraft}
+        >
+          {t("ReservationUnitEditor.saveAsDraft")}
+        </WhiteButton>
+        <WhiteButton
+          variant="primary"
+          // disabled={isSaving || !hasChanges}
```
Comment (joonatank) [inline apps/admin-ui/src/spa/ReservationUnit/edit/index.tsx:2212]: ???

**Review by vincit-matu — APPROVED** [review]
LGTM, after the changes we discovered while going through this are fixed

File: apps/admin-ui/src/spa/ReservationUnit/edit/form.ts [hunk 8]
```diff
@@ -0,0 +1,569 @@
...
+    hasReservationBegins: z.boolean(),
+    hasReservationEnds: z.boolean(),
+    hasBufferTimeBefore: z.boolean(),
+    hasBufferTimeAfter: z.boolean(),
+    hasCancellationRule: z.boolean(),
+  })
+  .superRefine((v, ctx) => {
+    if (v.isDraft || v.isArchived) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (joonatank) [reply apps/admin-ui/src/spa/ReservationUnit/edit/form.ts:227]: Not easily, so letting it be for now.

File: apps/admin-ui/src/spa/ReservationUnit/edit/index.tsx [hunk 9]
```diff
@@ -0,0 +1,2311 @@
...
+              >
+                <PricingType
+                  // TODO form index is bad, use pk or form key
+                  index={index}
+                  form={form}
+                  taxPercentageOptions={taxPercentageOptions}
+                />
+              </FieldGroup>
+            )
+        )}
```
[the remark this answers is not in the document]
Comment (joonatank) [reply apps/admin-ui/src/spa/ReservationUnit/edit/index.tsx:1308]: It works, it's just not that neat. So not changing it.

**Pull request merged**